File: impl/csharp/test/PrimitivesSmoke.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// Self-contained conformance smoke tests over the SDK primitives, anchored to independent
    /// standards vectors (FIPS 180-4 SHA-384, RFC 8949 canonical CBOR, the multiformats signer id,
    /// the §6.1 effect lattice, the twenty-channel registry, the §2.2 alg-level/profile floor). The
    /// authoritative cross-language grading is the naalp-conform harness against the shared corpus
    /// (== Go == Rust); these keep the published package independently checkable without the harness.
    /// </summary>
    public sealed class PrimitivesSmoke
    {
        private static byte[] Sha384(byte[] b)
        {
            using var sha = SHA384.Create();
            return sha.ComputeHash(b);
        }

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();

        [Fact]
        public void Sha384Kat()
        {
            Assert.Equal(
                "cb00753f45a35e8bb5a03d699ac65007272c32ab0eded1631a8b605a43ff5bed"
                + "8086072ba1e7cc2358baeca134c825a7",
                Hex(Sha384(Encoding.UTF8.GetBytes("abc"))));
        }

        [Fact]
        public void CborCanonicalEncodeAndContentId()
        {
            // canonical map: keys emitted in bytewise-ascending order regardless of input order
            var m = new Cbor.M(new List<Cbor.Pair>
            {
                new Cbor.Pair(new Cbor.U(3), new Cbor.U(4)),
                new Cbor.Pair(new Cbor.U(2), new Cbor.U(0)),
            });
            Assert.Equal("a2020003 04".Replace(" ", ""), Hex(Cbor.Encode(m)));
            byte[] cid = Cbor.ContentId(Cbor.Encode(m));
            Assert.Equal(0x20, cid[0]);
            Assert.Equal(0x30, cid[1]);
            Assert.Equal(2 + 48, cid.Length);
        }

        [Fact]
        public void CborRejectsNonCanonical()
        {
            // out-of-order / non-shortest / indefinite / duplicate-key
            foreach (string bad in new[] { "a202000100", "1800", "9f00ff", "a201000101" })
            {
                var ex = Assert.Throws<NaalpException>(() => Cbor.Decode(Convert.FromHexString(bad)));
                Assert.Equal("NonCanonical", ex.Kind);
            }
        }

        [Fact]
        public void CoseToBeSignedRfc9052()
        {
            byte[] tbs = Cose.ToBeSignedRaw(Convert.FromHexString("a1013830"), Convert.FromHexString("a10700"));
            // ["Signature1", ...]  ->  84 6a 5369676e6174757265 31
            Assert.StartsWith("846a5369676e617475726531", Hex(tbs));
        }

        [Fact]
        public void SignerIdForm()
        {
            byte[] pk = Cose.MldsaKeygen("ML-DSA-65", new byte[32]);
            Assert.StartsWith("b", Identity.SignerId(Cose.ALG_MLDSA65, pk));
        }

        [Fact]
        public void EffectLattice()
        {
            Assert.Equal(Policy.DESTRUCTIVE, Policy.NormalizeEffect(99)); // unknown -> destructive
            Assert.True(Policy.Authorizes(Policy.NON_IDEMPOTENT_WRITE, Policy.IDEMPOTENT_WRITE));
            Assert.False(Policy.Authorizes(Policy.READ_ONLY, Policy.DESTRUCTIVE));
        }

        [Fact]
        public void ChannelsRegistry()
        {
            Channels.KindSpec ks = Channels.Lookup(0x0004, 1); // Governance.Approval
            Assert.Equal("Approval", ks.Name);
            Assert.Equal((int)Policy.NON_IDEMPOTENT_WRITE, ks.Effect);
            Assert.False(ks.Variable);
            var ex = Assert.Throws<NaalpException>(() => Channels.Lookup(0x0000, 9999));
            Assert.Equal("UnknownKind", ex.Kind);
        }

        [Fact]
        public void RecordsDeterministic()
        {
            // a receipt body round-trips to stable bytes and the head is SHA-384 of it
            byte[] body = Records.ReceiptBody(new byte[48], Convert.FromHexString("2030" + new string('0', 96)), 0, 100);
            Assert.Equal(Hex(Sha384(body)), Hex(Records.ReceiptHead(body)));
        }

        [Fact]
        public void AlgLevelAndProfileFloor()
        {
            Assert.Equal(5, Cose.AlgLevel(Cose.ALG_MLDSA87).Level);
            Assert.Equal(3, Cose.AlgLevel(Cose.ALG_MLDSA65).Level);
            Assert.False(Cose.AlgLevel(999).Known);
            Assert.Equal(5, Cose.ProfileMinLevel(Cose.PROFILE_SOVEREIGN));
            Assert.Equal(3, Cose.ProfileMinLevel(Cose.PROFILE_PUBLIC));
        }
    }
}

File: impl/kotlin/src/test/kotlin/PrimitivesSmoke.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.security.MessageDigest

/**
 * Self-contained conformance smoke tests over the SDK primitives, anchored to independent
 * standards vectors (FIPS 180-4 SHA-384, RFC 8949 canonical CBOR, the multiformats signer id,
 * the §6.1 effect lattice, the twenty-channel registry). The authoritative cross-language grading
 * is the naalp-conform harness against the 239-case corpus; these keep the published package
 * independently checkable.
 *
 * Run (main()-driven, no test framework):
 *   java -cp "primitives-smoke.jar;<bcprov.jar>" sh.bubblefish.naalp.PrimitivesSmokeKt
 */

private class SmokeFailure(msg: String) : RuntimeException(msg)

private fun expect(what: String, cond: Boolean) {
    if (!cond) throw SmokeFailure(what)
    println("  ok  $what")
}

private fun sha384(b: ByteArray): ByteArray = MessageDigest.getInstance("SHA-384").digest(b)

fun main() {
    println("PrimitivesSmoke — standards-anchored primitives")

    // FIPS 180-4 SHA-384("abc")
    expect(
        "SHA-384(\"abc\")",
        Hex.encode(sha384("abc".toByteArray(Charsets.US_ASCII))) ==
            "cb00753f45a35e8bb5a03d699ac65007272c32ab0eded1631a8b605a43ff5bed8086072ba1e7cc2358baeca134c825a7",
    )

    // RFC 8949 canonical CBOR: keys emitted bytewise-ascending regardless of input order
    val m = Cbor.M(
        listOf(
            Cbor.Pair(Cbor.U(3), Cbor.U(4)),
            Cbor.Pair(Cbor.U(2), Cbor.U(0)),
        )
    )
    val enc = Cbor.encode(m)
    expect("canonical CBOR map order", Hex.encode(enc) == "a202000304")
    val cid = Cbor.contentId(enc)
    expect("content id multihash prefix", cid[0].toInt() == 0x20 && cid[1].toInt() == 0x30)
    expect("content id length", cid.size == 2 + 48)

    // strict decoder rejects non-canonical forms (out-of-order / non-shortest / indefinite / dup)
    for (bad in listOf("a202000100", "1800", "9f00ff", "a201000101")) {
        var kind = ""
        try {
            Cbor.decode(Hex.decode(bad))
        } catch (e: NaalpException) {
            kind = e.kind
        }
        expect("rejects non-canonical $bad", kind == "NonCanonical")
    }

    // RFC 9052 §4.4 Sig_structure begins with ["Signature1", ...]
    val tbs = Cose.toBeSignedRaw(Hex.decode("a1013830"), Hex.decode("a10700"))
    expect("ToBeSigned Sig_structure prefix", Hex.encode(tbs).startsWith("846a5369676e617475726531"))

    // multiformats self-certifying signer id: multibase base32 prefix 'b'
    val pk = Cose.mldsaKeygen("ML-DSA-65", ByteArray(32))
    val sid = Identity.signerId(Cose.ALG_MLDSA65, pk)
    expect("signer id multibase prefix", sid.startsWith("b"))

    // §6.1 effect lattice
    expect("effect: unknown -> destructive", Policy.normalizeEffect(99) == Policy.DESTRUCTIVE)
    expect("effect: niw authorizes iw", Policy.authorizes(Policy.NON_IDEMPOTENT_WRITE, Policy.IDEMPOTENT_WRITE))
    expect("effect: ro does not authorize destructive", !Policy.authorizes(Policy.READ_ONLY, Policy.DESTRUCTIVE))

    // twenty-channel registry: Governance.Approval (channel 4, kind 1)
    val ks = Channels.lookup(4, 1)
    expect("channel 4/1 = Approval", ks.name == "Approval")
    expect("channel 4/1 effect = non_idempotent_write", ks.effect.toLong() == Policy.NON_IDEMPOTENT_WRITE)
    expect("channel 4/1 not variable", !ks.variable)
    var unknownKind = ""
    try {
        Channels.lookup(0, 9999)
    } catch (e: NaalpException) {
        unknownKind = e.kind
    }
    expect("unknown channel/kind rejected", unknownKind.isNotEmpty())

    // record bodies are deterministic: receipt head == SHA-384(body)
    val body = Records.receiptBody(ByteArray(48), Hex.decode("2030" + "00".repeat(48)), 0, 100)
    expect("receipt head = SHA-384(body)", Records.receiptHead(body).contentEquals(sha384(body)))

    println("PrimitivesSmoke: PASS")
}
